File: leetcode/Ruby/No13.roman_to_integer.rb
=begin
Difficulty:
Easy

Desc:
Roman numerals are represented by seven different symbols: I, V, X, L, C, D and M.

Symbol       Value
I             1
V             5
X             10
L             50
C             100
D             500
M             1000
For example,
two is written as II in Roman numeral, just two one's added together.
Twelve is written as, XII, which is simply X + II.
The number twenty seven is written as XXVII, which is XX + V + II.

Roman numerals are usually written largest to smallest from left to right. However, the numeral for four is not IIII. Instead, the number four is written as IV.
Because the one is before the five we subtract it making four. The same principle applies to the number nine, which is written as IX.
There are six instances where subtraction is used:

I can be placed before V (5) and X (10) to make 4 and 9.
X can be placed before L (50) and C (100) to make 40 and 90.
C can be placed before D (500) and M (1000) to make 400 and 900.
Given a roman numeral, convert it to an integer. Input is guaranteed to be within the range from 1 to 3999.

Example:
Input: "III"
Output: 3

Input: "IV"
Output: 4

Input: "IX"
Output: 9

Input: "LVIII"
Output: 58
Explanation: C = 100, L = 50, XXX = 30 and III = 3.

Input: "MCMXCIV"
Output: 1994
Explanation: M = 1000, CM = 900, XC = 90 and IV = 4.
=end

# @param {String} s
# @return {Integer}
def roman_to_int(s)
  roman_map = { "I" => 1, "V" => 5, "X" => 10, "L" => 50, "C" => 100, "D" => 500, "M" => 1000 }
  result = 0
  i = s.size - 1
  while i >= 0 do
    str = s[i]
    num = roman_map[str]
    if i < s.size - 1 && num < roman_map[s[i + 1]] then
      result -= num
    else
      result += num
    end
    i -= 1
  end
  return result
end

# Test case
puts roman_to_int("MCMXCIV") # 1994
puts roman_to_int("LVIII") # 58
puts roman_to_int("III") # 3

File: leetcode/JavaScript/No13.roman-to-integer.js
/*
 * Difficulty:
 * Easy
 *
 * Desc:
 * Given a roman numeral, convert it to an integer.
 * Input is guaranteed to be within the range from 1 to 3999.
 *
 * Example:
 * roman -> MMMCCCXXXIII
 * num -> 3333
 *
 * 与第 12 题倒过来了，由罗马数字求出 int
 * 比较简单，只要找出罗马数字的规律即可：
 * 从末尾（右端）向左遍历，各个数字的左侧的罗马字母的值，如果小于当前数字，则需要做减法，否则做加法
 */

var ROMAN = {
  M: 1000,
  D: 500,
  C: 100,
  L: 50,
  X: 10,
  V: 5,
  I: 1
};

/**
 * @param {string} s
 * @return {number}
 */
var romanToInt = function(s) {
  let result = 0
  let i = s.length - 1

  while (i >= 0) {
    result += MAP[s[i]]

    let index = i - 1
    while (index >= 0 && MAP[s[index]] < MAP[s[i]]) {
      result -= MAP[s[index]]
      index -= 1
    }
    i = index
  }
  return result
}

romanToInt('MDCCC');
romanToInt('MCDXXXVII');
romanToInt('MCD');
romanToInt('MMMCMXCIX');
romanToInt('MMMCCCXXXIII');
romanToInt('M');
romanToInt('MM');
romanToInt('XIX');
romanToInt('I');